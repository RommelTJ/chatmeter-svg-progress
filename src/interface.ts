export interface ProgressProps {
  mode?: 'standard' | 'progress' | 'timed';
  duration?: number;
  begin?: number;
  repeatCount?: number | 'indefinite';
  from?: number;
  to?: number;
  progress?: number;
  revAnimationThreshold?: number;
  fireAnimationThreshold?: number;
}

export const STANDARD = 'standard';
export const PROGRESS = 'progress';
export const TIMED = 'timed';
export const INDEFINITE = 'indefinite';

export const DEFAULT_DURATION = 5; // 5 seconds
export const DEFAULT_BEGIN = 0; // 0 seconds
export const DEFAULT_FROM = 0;
export const DEFAULT_TO = 180;

export const defaultProps: Partial<ProgressProps> = {
  mode: STANDARD,
  duration: DEFAULT_DURATION,
  begin: DEFAULT_BEGIN,
  repeatCount: INDEFINITE,
  from: DEFAULT_FROM,
  to: DEFAULT_TO,
};
