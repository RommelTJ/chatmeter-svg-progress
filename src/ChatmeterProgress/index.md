---
nav:
  title: Components
  path: /components
---

## ChatmeterProgress

Demo:

```tsx
import React from 'react';
import { ChatmeterProgress } from 'chatmeter-svg-progress';

export default () => <ChatmeterProgress />;
```
