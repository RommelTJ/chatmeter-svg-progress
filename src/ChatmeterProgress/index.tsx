import React from 'react';
import { ProgressProps, defaultProps } from '../interface';

const ChatmeterProgress: React.FC<ProgressProps> = (props: ProgressProps) => {
  return <div>Hello World</div>;
};

ChatmeterProgress.defaultProps = defaultProps;
ChatmeterProgress.displayName = 'ChatmeterProgress';

export default ChatmeterProgress;
