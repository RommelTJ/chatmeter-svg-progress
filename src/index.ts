export { default as ChatmeterProgress } from './ChatmeterProgress';
