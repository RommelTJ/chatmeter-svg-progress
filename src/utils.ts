import { DEFAULT_DURATION } from './interface';

export const validatedProgress = (progress: number) => {
  if (progress < 0) return 0;
  if (progress > 100) return 100;
  return progress;
};

export const validateDuration = (duration: number | undefined) => {
  if (!duration) return DEFAULT_DURATION;
  if (duration < 0) return DEFAULT_DURATION;
  return duration;
};

export const validateThreshold = (
  duration: number,
  threshold: number | undefined,
) => {
  if (!threshold || threshold < 0) return DEFAULT_DURATION;
  return Math.min(duration, threshold);
};
