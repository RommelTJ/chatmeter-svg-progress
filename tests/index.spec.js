import React from 'react';
import { ChatmeterProgress } from '../src';
import {
  validateProgress,
  validateDuration,
  validateThreshold,
} from '../src/utils';
import { shallow } from 'enzyme';
import { describe, it } from '@jest/globals';
import { DEFAULT_DURATION } from '../src/interface';

describe('ChatmeterProgress', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<ChatmeterProgress />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should be titled chatmeter progress', () => {
    const wrapper = shallow(<ChatmeterProgress />);
    expect(wrapper.text()).toBe('Chatmeter Progress');
  });

  it('should support progress mode', () => {
    const wrapper = shallow(
      <ChatmeterProgress mode={'progress'} percent={50} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should support timed mode', () => {
    const wrapper = shallow(
      <ChatmeterProgress
        mode={'timed'}
        duration={10}
        revAnimationThreshold={8}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Utils', () => {
  it('should validate progress', () => {
    const normalProgress = 66;
    const validatedProgressNormal = validateProgress(normalProgress);
    expect(validatedProgressNormal).toBe(66);
  });

  it('should set negative progress to 0%', () => {
    const negativeProgress = -20;
    const validatedProgressNegative = validateProgress(negativeProgress);
    expect(validatedProgressNegative).toBe(0);
  });

  it('should set the max progress as 100%', () => {
    const extraProgress = 110;
    const validatedProgressExtra = validateProgress(extraProgress);
    expect(validatedProgressExtra).toBe(100);
  });

  it('should validate duration', () => {
    const duration = 10;
    const validatedDuration = validateDuration(duration);
    expect(validatedDuration).toBe(10);
  });

  it('should validate undefined duration', () => {
    const duration = undefined;
    const validatedDuration = validateDuration(duration);
    expect(validatedDuration).toBe(DEFAULT_DURATION);
  });

  it('should validate negative duration', () => {
    const duration = -10;
    const validatedDuration = validateDuration(duration);
    expect(validatedDuration).toBe(DEFAULT_DURATION);
  });

  it('should validate threshold', () => {
    const duration = DEFAULT_DURATION;
    const threshold = 3;
    const validatedThreshold = validateThreshold(duration, threshold);
    expect(validatedThreshold).toBe(3);
  });

  it('should return the min of two valid values for threshold', () => {
    const duration = 7;
    const threshold = 18;
    const validatedThreshold = validateThreshold(duration, threshold);
    expect(validatedThreshold).toBe(7);
  });

  it('should sanitize for negative thresholds', () => {
    const duration = 20;
    const threshold = -10;
    const validatedThreshold = validateThreshold(duration, threshold);
    expect(validatedThreshold).toBe(DEFAULT_DURATION);
  });
});
