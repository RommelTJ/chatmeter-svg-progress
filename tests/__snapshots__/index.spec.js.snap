// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatmeterProgress renders correctly 1`] = `
<svg
  height="300px"
  viewBox="0 0 300 300"
  width="300px"
>
  <title>
    Chatmeter Progress
  </title>
  <defs>
    <linearGradient
      id="linearGradient-1"
      x1="61.2218064%"
      x2="-189.036108%"
      y1="45.4857754%"
      y2="89.4496267%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-2"
      x1="162.673729%"
      x2="-278.27306%"
      y1="25.8979885%"
      y2="115.014368%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-3"
      x1="216.032141%"
      x2="-149.704183%"
      y1="5.80605312%"
      y2="107.967881%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-4"
      x1="325.353138%"
      x2="-143.462461%"
      y1="-32.2323462%"
      y2="109.083144%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-5"
      x1="616.631353%"
      x2="-239.875432%"
      y1="-91.2613%"
      y2="122.384847%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-6"
      x1="1278.73885%"
      x2="-451.537478%"
      y1="-197.268908%"
      y2="150.210084%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
  </defs>
  <g
    fill="none"
    fillRule="evenodd"
    stroke="none"
    strokeWidth="1"
  >
    <path
      d="M58.8909951,148.32676 L58.7209951,148.38676 L27.1309951,159.29676 C25.1896347,159.900496 23.3902921,160.890304 21.8409951,162.20676 C19.810939,164.153597 18.6625281,166.844056 18.6609651,169.65676 C18.657259,173.13756 20.3869026,176.391725 23.2741225,178.335934 C26.1613425,180.280142 29.8270779,180.659141 33.0509951,179.34676 L33.1109951,179.34676 L33.4609951,179.17676 L159.130995,113.70676 L58.8909951,148.32676 Z"
      fill="#E4670D"
      fillRule="nonzero"
      id="Needle"
      transform="translate(88.895980, 146.912391) rotate(-156.000000) translate(-88.895980, -146.912391)"
    />
    <animateTransform
      attributeName="transform"
      begin="0s"
      dur="5s"
      from="0 150 150"
      repeatCount="indefinite"
      to="180 150 150"
      type="rotate"
    />
  </g>
  <g>
    <path
      d="M195.14,63.19 L162.25,116.67 L162.25,116.67 C177.443192,124.147995 190.331427,135.595374 199.55,149.8 L261.12,105.51 C243.761035,85.2286255 220.812806,70.5094828 195.14,63.19 Z"
      fill="url(#linearGradient-1)"
      fillRule="nonzero"
    />
    <path
      d="M157.99,58 C149.642736,58.011491 141.314255,58.7914492 133.11,60.33 L126.11,107.27 L126.11,107.27 C136.162534,107.806247 146.061319,109.971079 155.42,113.68 L181.36,60.05 C173.643381,58.696388 165.824435,58.0105156 157.99,58 Z"
      fill="url(#linearGradient-2)"
      fillRule="nonzero"
    />
    <path
      d="M91.84,111.78 C100.79359,108.847388 110.139364,107.286387 119.56,107.15 L120.79,63.21 C108.489406,66.7079437 96.7485952,71.9369312 85.92,78.74 L91.84,111.78 Z"
      fill="url(#linearGradient-3)"
      fillRule="nonzero"
    />
    <path
      d="M60.44,99.41 L71.54,121.26 L71.54,121.26 C75.46259,118.802669 79.5641142,116.643269 83.81,114.8 L75.29,86.14 C70.0258203,90.1986546 65.0628117,94.6336125 60.44,99.41 L60.44,99.41 Z"
      fill="url(#linearGradient-4)"
      fillRule="nonzero"
    />
    <path
      d="M46.71,116.06 L58.96,130.67 L58.96,130.67 C61.03,128.83 63.17,127.08 65.39,125.44 L53.2,107.44 C50.93,110.22 48.78,113.11 46.71,116.06 Z"
      fill="url(#linearGradient-5)"
      fillRule="nonzero"
    />
    <path
      d="M39,128.42 L50.54,139.18 C51.42,138.18 52.33,137.18 53.26,136.18 L41.02,124.9 C40.33,126.06 39.64,127.23 39,128.42 Z"
      fill="url(#linearGradient-6)"
      fillRule="nonzero"
    />
  </g>
</svg>
`;

exports[`ChatmeterProgress should support progress mode 1`] = `
<svg
  height="300px"
  viewBox="0 0 300 300"
  width="300px"
>
  <title>
    Chatmeter Progress
  </title>
  <defs>
    <linearGradient
      id="linearGradient-1"
      x1="61.2218064%"
      x2="-189.036108%"
      y1="45.4857754%"
      y2="89.4496267%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-2"
      x1="162.673729%"
      x2="-278.27306%"
      y1="25.8979885%"
      y2="115.014368%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-3"
      x1="216.032141%"
      x2="-149.704183%"
      y1="5.80605312%"
      y2="107.967881%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-4"
      x1="325.353138%"
      x2="-143.462461%"
      y1="-32.2323462%"
      y2="109.083144%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-5"
      x1="616.631353%"
      x2="-239.875432%"
      y1="-91.2613%"
      y2="122.384847%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-6"
      x1="1278.73885%"
      x2="-451.537478%"
      y1="-197.268908%"
      y2="150.210084%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
  </defs>
  <g
    fill="none"
    fillRule="evenodd"
    stroke="none"
    strokeWidth="1"
  >
    <path
      d="M58.8909951,148.32676 L58.7209951,148.38676 L27.1309951,159.29676 C25.1896347,159.900496 23.3902921,160.890304 21.8409951,162.20676 C19.810939,164.153597 18.6625281,166.844056 18.6609651,169.65676 C18.657259,173.13756 20.3869026,176.391725 23.2741225,178.335934 C26.1613425,180.280142 29.8270779,180.659141 33.0509951,179.34676 L33.1109951,179.34676 L33.4609951,179.17676 L159.130995,113.70676 L58.8909951,148.32676 Z"
      fill="#E4670D"
      fillRule="nonzero"
      id="Needle"
      transform="translate(88.895980, 146.912391) rotate(-156.000000) translate(-88.895980, -146.912391)"
    />
    <animateTransform
      attributeName="transform"
      begin="0s"
      dur="0s"
      from="0 150 150"
      repeatCount={0}
      to="0 150 150"
      type="rotate"
    />
  </g>
  <g>
    <path
      d="M195.14,63.19 L162.25,116.67 L162.25,116.67 C177.443192,124.147995 190.331427,135.595374 199.55,149.8 L261.12,105.51 C243.761035,85.2286255 220.812806,70.5094828 195.14,63.19 Z"
      fill="url(#linearGradient-1)"
      fillRule="nonzero"
    />
    <path
      d="M157.99,58 C149.642736,58.011491 141.314255,58.7914492 133.11,60.33 L126.11,107.27 L126.11,107.27 C136.162534,107.806247 146.061319,109.971079 155.42,113.68 L181.36,60.05 C173.643381,58.696388 165.824435,58.0105156 157.99,58 Z"
      fill="url(#linearGradient-2)"
      fillRule="nonzero"
    />
    <path
      d="M91.84,111.78 C100.79359,108.847388 110.139364,107.286387 119.56,107.15 L120.79,63.21 C108.489406,66.7079437 96.7485952,71.9369312 85.92,78.74 L91.84,111.78 Z"
      fill="url(#linearGradient-3)"
      fillRule="nonzero"
    />
    <path
      d="M60.44,99.41 L71.54,121.26 L71.54,121.26 C75.46259,118.802669 79.5641142,116.643269 83.81,114.8 L75.29,86.14 C70.0258203,90.1986546 65.0628117,94.6336125 60.44,99.41 L60.44,99.41 Z"
      fill="url(#linearGradient-4)"
      fillRule="nonzero"
    />
    <path
      d="M46.71,116.06 L58.96,130.67 L58.96,130.67 C61.03,128.83 63.17,127.08 65.39,125.44 L53.2,107.44 C50.93,110.22 48.78,113.11 46.71,116.06 Z"
      fill="url(#linearGradient-5)"
      fillRule="nonzero"
    />
    <path
      d="M39,128.42 L50.54,139.18 C51.42,138.18 52.33,137.18 53.26,136.18 L41.02,124.9 C40.33,126.06 39.64,127.23 39,128.42 Z"
      fill="url(#linearGradient-6)"
      fillRule="nonzero"
    />
  </g>
</svg>
`;

exports[`ChatmeterProgress should support timed mode 1`] = `
<svg
  height="300px"
  viewBox="0 0 300 300"
  width="300px"
>
  <title>
    Chatmeter Progress
  </title>
  <defs>
    <linearGradient
      id="linearGradient-1"
      x1="61.2218064%"
      x2="-189.036108%"
      y1="45.4857754%"
      y2="89.4496267%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-2"
      x1="162.673729%"
      x2="-278.27306%"
      y1="25.8979885%"
      y2="115.014368%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-3"
      x1="216.032141%"
      x2="-149.704183%"
      y1="5.80605312%"
      y2="107.967881%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-4"
      x1="325.353138%"
      x2="-143.462461%"
      y1="-32.2323462%"
      y2="109.083144%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-5"
      x1="616.631353%"
      x2="-239.875432%"
      y1="-91.2613%"
      y2="122.384847%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
    <linearGradient
      id="linearGradient-6"
      x1="1278.73885%"
      x2="-451.537478%"
      y1="-197.268908%"
      y2="150.210084%"
    >
      <stop
        offset="0%"
        stopColor="#16365B"
      />
      <stop
        offset="100%"
        stopColor="#16365B"
        stopOpacity="0.1"
      />
    </linearGradient>
  </defs>
  <g
    fill="none"
    fillRule="evenodd"
    stroke="none"
    strokeWidth="1"
  >
    <path
      d="M58.8909951,148.32676 L58.7209951,148.38676 L27.1309951,159.29676 C25.1896347,159.900496 23.3902921,160.890304 21.8409951,162.20676 C19.810939,164.153597 18.6625281,166.844056 18.6609651,169.65676 C18.657259,173.13756 20.3869026,176.391725 23.2741225,178.335934 C26.1613425,180.280142 29.8270779,180.659141 33.0509951,179.34676 L33.1109951,179.34676 L33.4609951,179.17676 L159.130995,113.70676 L58.8909951,148.32676 Z"
      fill="#E4670D"
      fillRule="nonzero"
      id="Needle"
      transform="translate(88.895980, 146.912391) rotate(-156.000000) translate(-88.895980, -146.912391)"
    />
    <animateTransform
      attributeName="transform"
      begin="0s"
      dur="10s"
      from="0 150 150"
      repeatCount="indefinite"
      to="180 150 150"
      type="rotate"
    />
  </g>
  <g>
    <path
      d="M195.14,63.19 L162.25,116.67 L162.25,116.67 C177.443192,124.147995 190.331427,135.595374 199.55,149.8 L261.12,105.51 C243.761035,85.2286255 220.812806,70.5094828 195.14,63.19 Z"
      fill="url(#linearGradient-1)"
      fillRule="nonzero"
    />
    <path
      d="M157.99,58 C149.642736,58.011491 141.314255,58.7914492 133.11,60.33 L126.11,107.27 L126.11,107.27 C136.162534,107.806247 146.061319,109.971079 155.42,113.68 L181.36,60.05 C173.643381,58.696388 165.824435,58.0105156 157.99,58 Z"
      fill="url(#linearGradient-2)"
      fillRule="nonzero"
    />
    <path
      d="M91.84,111.78 C100.79359,108.847388 110.139364,107.286387 119.56,107.15 L120.79,63.21 C108.489406,66.7079437 96.7485952,71.9369312 85.92,78.74 L91.84,111.78 Z"
      fill="url(#linearGradient-3)"
      fillRule="nonzero"
    />
    <path
      d="M60.44,99.41 L71.54,121.26 L71.54,121.26 C75.46259,118.802669 79.5641142,116.643269 83.81,114.8 L75.29,86.14 C70.0258203,90.1986546 65.0628117,94.6336125 60.44,99.41 L60.44,99.41 Z"
      fill="url(#linearGradient-4)"
      fillRule="nonzero"
    />
    <path
      d="M46.71,116.06 L58.96,130.67 L58.96,130.67 C61.03,128.83 63.17,127.08 65.39,125.44 L53.2,107.44 C50.93,110.22 48.78,113.11 46.71,116.06 Z"
      fill="url(#linearGradient-5)"
      fillRule="nonzero"
    />
    <path
      d="M39,128.42 L50.54,139.18 C51.42,138.18 52.33,137.18 53.26,136.18 L41.02,124.9 C40.33,126.06 39.64,127.23 39,128.42 Z"
      fill="url(#linearGradient-6)"
      fillRule="nonzero"
    />
  </g>
</svg>
`;
